src/tree.rs

```diff
@@ -415,35 +435,73 @@ where
         let mut tree_cache = TreeCache::new(self.reader, first_version)?;
         for (idx, value_set) in value_sets.into_iter().enumerate() {
             let version = first_version + idx as u64;
-            value_set
-                .into_iter()
-                .enumerate()
-                .try_for_each(|(i, (key, value))| {
-                    let action = if value.is_some() { "insert" } else { "delete" };
-                    let value_hash = value.as_ref().map(|v| ValueHash::with::<H>(v));
-                    tree_cache.put_value(version, key, value);
-                    self.put(key, value_hash, version, &mut tree_cache)
-                        .with_context(|| {
-                            format!(
-                                "failed to {} key {} for version {}, key = {:?}",
-                                action, i, version, key
-                            )
-                        })
-                })?;
+            for (i, (key, value)) in value_set.into_iter().enumerate() {
+                let action = if value.is_some() { "insert" } else { "delete" };
+                let value_hash = value.as_ref().map(|v| ValueHash::with::<H>(v));
+                tree_cache.put_value(version, key, value);
+                self.put(key, value_hash, version, &mut tree_cache, false)
+                    .with_context(|| {
+                        format!(
+                            "failed to {} key {} for version {}, key = {:?}",
+                            action, i, version, key
+                        )
+                    })?;
+            }
+
             // Freezes the current cache to make all contents in the current cache immutable.
             tree_cache.freeze()?;
         }
 
         Ok(tree_cache.into())
     }
 
+    /// Same as [`put_value_sets`], this method returns a Merkle proof for every update of the Merkle tree.
+    /// The proofs can be verified using the [`verify_update`] method, which requires the old `root_hash`, the `merkle_proof` and the new `root_hash`
+    pub fn put_value_sets_with_proof(
```

Comment:
I don't know if it would be clearer to be honest, specially because we do populate values of a proof vector on each iteration of the nested loops in the proof generating version